src/Microsoft.Azure.SignalR.Common/ServiceCommunicationScope.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+
+namespace Microsoft.Azure.SignalR.Common
+{
+    /// <summary>
+    /// Represents a generic nestable scope for grouping operations with the same service communication policy.
+    /// Derived classes must ensure that the service protocol can understand and transmit this policy to the service.
+    /// </summary>
+    /// The following is ensured for all send operations to the service within the scope:
+    /// -  the same service connection is used in ServiceTransportType.Persistent mode
+    /// -  the same service connection is used for calls outside of hub call context
+    /// -  only the value of the most nested scope will flow along with the messages
+    /// 
+    /// <typeparam name="TNestableProps">Type representing policy to flow</typeparam>
+    public class ServiceCommunicationScope<TNestableProps> : IDisposable
+    {
+        private ScopePropertiesAccessor<TNestableProps> _previousScope;
+        private ClientConnectionScopeInternal _serviceConnectionScope;
+
+        protected ServiceCommunicationScope(TNestableProps properties)
+        {
+            _previousScope = ScopePropertiesAccessor<TNestableProps>.Current;
+            ScopePropertiesAccessor<TNestableProps>.Current = new ScopePropertiesAccessor<TNestableProps>() { Properties = properties };
+            _serviceConnectionScope = new ClientConnectionScopeInternal();
```

Comment:
looks like the logic in `ClientConnectionScopeInternal`'s ctor is to manipulate `ScopePropertiesAccessor<TNestableProps>.Current`? how about moving the logic into `Current`'s `set` logic and we can remove this class `ClientConnectionScopeInternal`?

Reply:
Hmm... and use something like this:
~~~
try {
  ScopePropertiesAccessor<ClientConnectionScopeProperties>.Current = new ClientConnectionScopeProperties();
  <scope work>
}
finally
{
  ScopePropertiesAccessor<ClientConnectionScopeProperties>.Current = null;
}
~~~
I would not mind but `ClientConnectionScopeInternal` is being used in 2 places and both places would have to take care of correctly handling instantiating/cleaning up the async local (e.g. lines 40-42, 53-57). This class seems like a nice wrapper to handle this logic.

Reply:
The implementation of this class looks a little fragile to me that it involves both `ClientConnectionScopeInternal` and `ScopePropertiesAccessor<TNestableProps>.Current` while inside `ClientConnectionScopeInternal` it also manipulates `ScopePropertiesAccessor<TNestableProps>.Current` and the order of calling these two in `ctor` and `dispose` I believe matters to make it work. Why it goes this way depends on how `ClientConnectionScopeInternal` is implemented internally while I don't quite get it why `ServiceCommunicationScope` has something to do with `ClientConnectionScopeInternal`. Maybe directly manipulate `ScopePropertiesAccessor` instead of depending on `ClientConnectionScopeInternal` can make the logic clearer?